File: 0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        /*ListNode* temp = head ;
        map<ListNode* , int> mpp ;
        while(temp != NULL){
            if(mpp.find(temp) != mpp.end())
                return 1 ;
            mpp[temp] = 1 ;
            temp = temp -> next ;
        }
        return 0 ;  */ 
        ListNode* slow = head ;
        ListNode* fast = head ;
        while(fast != NULL && fast -> next != NULL){
            slow = slow -> next ;
            fast = fast -> next -> next ;
            if(fast == slow )
                return 1 ;
        }
        return 0 ;
    }
};

File: 0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if(head == null)    return false ;    
        HashMap<ListNode,Integer> m = new HashMap<>() ;
        ListNode temp = head ;
        while(temp != null){
            if(m.containsKey(temp)){
                return true ;
            }
            else
                m.put(temp,1);
            temp = temp.next ;
        }
        return false ;
    }
}